File: src/solar/util/archive/GlobalArchiveUtil.ts
/** @noSelf **/
import CodecUtil from "../math/CodecUtil";

/**
 将存档key 与存档值 作为Json字符串 统一存储到平台服务器上
 因此请尽量减少key的长度 以多存内容

 需要手动在作者之家后台 全局存档 创建可读可写的 _SLGA_0 -> _SLGA_15 key值
 @deprecated 正在测试中...
 */
export default class GlobalArchiveUtil {
    //加密 请在地图发布前就确定是否需要加密以及加密的方式。发布后不能更改此选项 否则会造成旧存档被清空
    static encrypt: boolean = true;
    static ArchiveKeyCountMax: number = 15;
    static salt: string = "s_l_g_a_z_b_s_q_j";
    static baseArchiveKeyPrefix: string = "_SLGA_";
    //
    static data: { [key: string]: any } = {}


    /**
     *
     * @param key
     */
    static get(key: string) {
        GlobalArchiveUtil._init0();
        return GlobalArchiveUtil.data[key];
    }

    static set(key: string, value: any) {
        GlobalArchiveUtil._init0();
        GlobalArchiveUtil.data[key] = value;
        GlobalArchiveUtil.save()
    }

    /**
     *获取所有存档
     */
    static getAllArchive(): any {
        GlobalArchiveUtil._init0();
        return GlobalArchiveUtil.data;
    }

    /**
     * 打印所有存档 调试时用
     * @param whichPlayer
     */
    static printAllArchive(): void {
        GlobalArchiveUtil._init0();
        let data = GlobalArchiveUtil.data;
        if (!data) {
            return;
        }
        print("========打印全局存档 开始========")
        for (let dataKey in data) {
            print(dataKey + "=[" + data[dataKey] + "]")
        }
        print("========打印全局存档 结束========")
    }

    /**
     * base
     * 清空所有全局存档
     */
    static clearAllArchive(): void {
        for (let i = 0; i < GlobalArchiveUtil.ArchiveKeyCountMax; i++) {
            let key: string = GlobalArchiveUtil.baseArchiveKeyPrefix + i;
            DzAPI_Map_Global_StoreString(key, "");
        }
        print("已清理全部全局存档!")
    }

    /**
     * 保存所有全局存档数据
     */
    static save(): void {
        let data = GlobalArchiveUtil.data;
        let dataJsonStr: string = JSON.stringify(data);
        if (GlobalArchiveUtil.encrypt) {
            dataJsonStr = CodecUtil.saesEncode(dataJsonStr, gameName + GlobalArchiveUtil.salt)
        }
        let dataJsonStrLength = dataJsonStr.length;
        let keyLength = math.floor(dataJsonStrLength / 64) + 1;
        log.debug("全局存档栏位大小为:" + keyLength)
        if (keyLength > GlobalArchiveUtil.ArchiveKeyCountMax) {
            log.errorWithTraceBack("全局栏位大小超过" + GlobalArchiveUtil.ArchiveKeyCountMax + "！" + keyLength)
            return;
        }
        for (let i = 0; i < keyLength; i++) {
            let endIndex = (i + 1) * 64;
            if (endIndex > dataJsonStrLength) {
                endIndex = dataJsonStrLength;
            }
            let oneData = dataJsonStr.substring(i * 64, endIndex)
            let key: string = GlobalArchiveUtil.baseArchiveKeyPrefix + i;
            DzAPI_Map_Global_StoreString(key, oneData);
        }
        //清空后一个存档栏位 以标记结束
        let key: string = GlobalArchiveUtil.baseArchiveKeyPrefix + keyLength;
        if (keyLength < GlobalArchiveUtil.ArchiveKeyCountMax) {
            DzAPI_Map_Global_StoreString(key, "");
        }

    }


    /**
     * 初始化存档数据
     * @private
     */
    private static isInit = false;

    private static _init0(): boolean {
        if (GlobalArchiveUtil.isInit) {
            return true;
        }
        GlobalArchiveUtil.isInit = true;
        let data: any = {};
        let dataJsonStr: string = "";
        for (let i = 0; i < GlobalArchiveUtil.ArchiveKeyCountMax; i++) {
            let key: string = GlobalArchiveUtil.baseArchiveKeyPrefix + i;
            let serverValue: string = DzAPI_Map_Global_GetStoreString(key);
            if (serverValue && serverValue.length > 0) {
                dataJsonStr = dataJsonStr + serverValue;
            } else {
                break;
            }
        }
        //转json了
        if (dataJsonStr.length < 2) {
            return true;
        }
        try {
            if (GlobalArchiveUtil.encrypt) {
                dataJsonStr = CodecUtil.saesDecode(dataJsonStr, gameName + GlobalArchiveUtil.salt)
            }
            data = JSON.parse(dataJsonStr);
            log.debug("全局存档数据为:" + dataJsonStr)
            //设置
            GlobalArchiveUtil.data = data;
            return true;
        } catch (e) {
            log.errorWithTraceBack("全局存档Json解析出错!")
            DisplayTimedTextToPlayer(GetLocalPlayer(), 0, 0, 60, '全局存档Json解析出错！');
            print("需要解析的Json字符串为[" + dataJsonStr + "]")
        }
        return true;
    }
}

File: scripts/solar/util/archive/GlobalArchiveUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__StringSubstring = ____lualib.__TS__StringSubstring
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["7"] = 2,["8"] = 2,["9"] = 11,["10"] = 11,["11"] = 11,["13"] = 11,["14"] = 25,["15"] = 26,["16"] = 27,["17"] = 25,["18"] = 30,["19"] = 31,["20"] = 32,["21"] = 33,["22"] = 30,["23"] = 39,["24"] = 40,["25"] = 41,["26"] = 39,["27"] = 48,["28"] = 49,["29"] = 50,["30"] = 51,["33"] = 54,["34"] = 55,["35"] = 56,["37"] = 58,["38"] = 48,["39"] = 65,["41"] = 66,["42"] = 66,["43"] = 67,["44"] = 68,["45"] = 66,["48"] = 70,["49"] = 65,["50"] = 76,["51"] = 77,["52"] = 78,["53"] = 79,["54"] = 80,["56"] = 82,["57"] = 83,["58"] = 84,["59"] = 85,["60"] = 86,["64"] = 89,["65"] = 89,["66"] = 90,["67"] = 91,["68"] = 92,["70"] = 94,["71"] = 95,["72"] = 96,["73"] = 89,["76"] = 99,["77"] = 100,["78"] = 101,["80"] = 76,["81"] = 113,["82"] = 114,["83"] = 115,["85"] = 117,["86"] = 118,["87"] = 119,["89"] = 120,["90"] = 120,["91"] = 121,["92"] = 122,["93"] = 123,["94"] = 124,["98"] = 120,["101"] = 130,["102"] = 131,["106"] = 143,["107"] = 144,["108"] = 144,["109"] = 144,["110"] = 144,["111"] = 144,["112"] = 144,["113"] = 144,["114"] = 145,["117"] = 134,["118"] = 135,["120"] = 137,["121"] = 138,["122"] = 140,["123"] = 141,["129"] = 133,["132"] = 147,["133"] = 113,["134"] = 13,["135"] = 14,["136"] = 15,["137"] = 16,["138"] = 18,["139"] = 111,["140"] = 11});
local ____exports = {}
local ____CodecUtil = require("solar.util.math.CodecUtil")
local CodecUtil = ____CodecUtil.default
____exports.default = __TS__Class()
local GlobalArchiveUtil = ____exports.default
GlobalArchiveUtil.name = "GlobalArchiveUtil"
function GlobalArchiveUtil.prototype.____constructor(self)
end
function GlobalArchiveUtil.get(self, key)
    ____exports.default:_init0()
    return ____exports.default.data[key]
end
function GlobalArchiveUtil.set(self, key, value)
    ____exports.default:_init0()
    ____exports.default.data[key] = value
    ____exports.default:save()
end
function GlobalArchiveUtil.getAllArchive(self)
    ____exports.default:_init0()
    return ____exports.default.data
end
function GlobalArchiveUtil.printAllArchive(self)
    ____exports.default:_init0()
    local data = ____exports.default.data
    if not data then
        return
    end
    print("========打印全局存档 开始========")
    for dataKey in pairs(data) do
        print(((dataKey .. "=[") .. tostring(data[dataKey])) .. "]")
    end
    print("========打印全局存档 结束========")
end
function GlobalArchiveUtil.clearAllArchive(self)
    do
        local i = 0
        while i < ____exports.default.ArchiveKeyCountMax do
            local key = ____exports.default.baseArchiveKeyPrefix .. tostring(i)
            DzAPI_Map_Global_StoreString(key, "")
            i = i + 1
        end
    end
    print("已清理全部全局存档!")
end
function GlobalArchiveUtil.save(self)
    local data = ____exports.default.data
    local dataJsonStr = JSON:stringify(data)
    if ____exports.default.encrypt then
        dataJsonStr = CodecUtil:saesEncode(dataJsonStr, gameName .. ____exports.default.salt)
    end
    local dataJsonStrLength = #dataJsonStr
    local keyLength = math.floor(dataJsonStrLength / 64) + 1
    log.debug("全局存档栏位大小为:" .. tostring(keyLength))
    if keyLength > ____exports.default.ArchiveKeyCountMax then
        log.errorWithTraceBack((("全局栏位大小超过" .. tostring(____exports.default.ArchiveKeyCountMax)) .. "！") .. tostring(keyLength))
        return
    end
    do
        local i = 0
        while i < keyLength do
            local endIndex = (i + 1) * 64
            if endIndex > dataJsonStrLength then
                endIndex = dataJsonStrLength
            end
            local oneData = __TS__StringSubstring(dataJsonStr, i * 64, endIndex)
            local key = ____exports.default.baseArchiveKeyPrefix .. tostring(i)
            DzAPI_Map_Global_StoreString(key, oneData)
            i = i + 1
        end
    end
    local key = ____exports.default.baseArchiveKeyPrefix .. tostring(keyLength)
    if keyLength < ____exports.default.ArchiveKeyCountMax then
        DzAPI_Map_Global_StoreString(key, "")
    end
end
function GlobalArchiveUtil._init0(self)
    if ____exports.default.isInit then
        return true
    end
    ____exports.default.isInit = true
    local data = {}
    local dataJsonStr = ""
    do
        local i = 0
        while i < ____exports.default.ArchiveKeyCountMax do
            local key = ____exports.default.baseArchiveKeyPrefix .. tostring(i)
            local serverValue = DzAPI_Map_Global_GetStoreString(key)
            if serverValue and #serverValue > 0 then
                dataJsonStr = dataJsonStr .. serverValue
            else
                break
            end
            i = i + 1
        end
    end
    if #dataJsonStr < 2 then
        return true
    end
    do
        local function ____catch(e)
            log.errorWithTraceBack("全局存档Json解析出错!")
            DisplayTimedTextToPlayer(
                GetLocalPlayer(),
                0,
                0,
                60,
                "全局存档Json解析出错！"
            )
            print(("需要解析的Json字符串为[" .. dataJsonStr) .. "]")
        end
        local ____try, ____hasReturned, ____returnValue = pcall(function()
            if ____exports.default.encrypt then
                dataJsonStr = CodecUtil:saesDecode(dataJsonStr, gameName .. ____exports.default.salt)
            end
            data = JSON:parse(dataJsonStr)
            log.debug("全局存档数据为:" .. dataJsonStr)
            ____exports.default.data = data
            return true, true
        end)
        if not ____try then
            ____hasReturned, ____returnValue = ____catch(____hasReturned)
        end
        if ____hasReturned then
            return ____returnValue
        end
    end
    return true
end
GlobalArchiveUtil.encrypt = true
GlobalArchiveUtil.ArchiveKeyCountMax = 15
GlobalArchiveUtil.salt = "s_l_g_a_z_b_s_q_j"
GlobalArchiveUtil.baseArchiveKeyPrefix = "_SLGA_"
GlobalArchiveUtil.data = {}
GlobalArchiveUtil.isInit = false
____exports.default = GlobalArchiveUtil
return ____exports
